### PR TITLE
Feat: add the ability to add in additional route properties

### PR DESCRIPTION
**Pull request by barelyhuman**
# Changes 
ability to add and modify route properties to modify how the routes might behave. One example is to be able to create a schema split which https://github.com/nearform/mercurius-dynamic-schema/ tries to achieve. 

## Reasoning
Helps close https://github.com/nearform/mercurius-dynamic-schema/issues/4

### CODE REVIEW

**Comment by simoneb** [comment]
For context, we tried various solutions in https://github.com/nearform/mercurius-dynamic-schema to extend core's functionality to achieve what we needed, and eventually realized that with this simple change in mercurius we'd have to get rid of unnecessary complications (and duplications) in our core. Hence we're proposing this change.

File: lib/routes.js [hunk 1]
```diff
@@ -200,7 +200,8 @@ module.exports = async function (app, opts) {
     persistedQueryProvider,
     allowBatchedQueries,
     keepAlive,
-    fullWsTransport
+    fullWsTransport,
+    additionalRouteProps
```
Comment (jonnydgreen) [inline lib/routes.js:204]: Hello! Worth adding some docs and types for this do you think?

File: index.js [hunk 2]
```diff
@@ -210,7 +210,8 @@ const mercurius = fp(async function (app, opts) {
       entityResolversFactory: undefined,
       subscriptionContextFn,
       keepAlive,
-      fullWsTransport
+      fullWsTransport,
+      additionalRouteProps: opts.additionalRouteProps
```
Comment (jonnydgreen) [inline index.js:214]: Hello! Worth adding some docs and types for this do you think?

Comment (simoneb) [reply index.js:214]: Yes indeed

**Review by mcollina — CHANGES_REQUESTED** [review]
can you add docs and tests?

**Comment by barelyhuman** [comment]
> can you add docs and tests?

Will do

File: test/routes.js [hunk 3]
```diff
@@ -2187,3 +2187,100 @@ test('if ide has plugin set, serve config.js with the correct endpoint', async (
...
+
+test('GET graphql endpoint, additionalRouteProps should ignore custom handler', async (t) => {
+  const app = Fastify()
+  const schema = `
+    type Query {
+      add(x: Int, y: Int): Int
+    }
+  `
+
+  let executed = false
```
Comment (simoneb) [inline test/routes.js:2259]: for this type of async tests, using t.plan is preferred. in my opinion this is also the only case in which you want to use t.plan, but in such cases, I prefer to use it versus setting a flag like you did

File: test/routes.js [hunk 4]
```diff
@@ -2187,3 +2187,100 @@ test('if ide has plugin set, serve config.js with the correct endpoint', async (
...
+
+  app.register(GQL, {
+    schema,
+    additionalRouteProps: {
+      // @ts-ignore
```
Comment (simoneb) [inline test/routes.js:2264]: why this ignore?

File: lib/routes.js [hunk 5]
```diff
@@ -214,6 +215,12 @@ module.exports = async function (app, opts) {
     notSupportedError
   } = persistedQueryProvider || {}
 
+  const normalizedRouteProps = { ...additionalRouteProps }
+  if (normalizedRouteProps.handler || normalizedRouteProps.wsHandler) {
+    normalizedRouteProps.handler = undefined
+    normalizedRouteProps.wsHandler = undefined
+  }
```
Comment (simoneb) [inline lib/routes.js:222]: why this change?

**Review by simoneb — CHANGES_REQUESTED** [review]
See inline comments. Plus, we should also include this in the documentation https://mercurius.dev/#/docs/api/options?id=mercurius

File: test/routes.js (hunk 4 above)
Comment (barelyhuman) [reply test/routes.js:2264]: The types omit it so added it for my editors tslint, accidentally committed it 🤦‍♂️

File: lib/routes.js (hunk 5 above)
Comment (barelyhuman) [reply lib/routes.js:222]: The post handler adds a callback in which case fastify errors saying that there's 2 handlers. so normalised it to avoid even considering the handler provided from the options

File: test/routes.js (hunk 3 above)
Comment (barelyhuman) [reply test/routes.js:2259]: wouldn't a t.plan wait for the handler to execute? 

i don't want it to execute here , hence the flag

Comment (simoneb) [reply test/routes.js:2259]: t.plan would cause the test runner to wait to finish the test until all planned assertions have run, which is what you're trying to accomplish here

Comment (simoneb) [reply test/routes.js:2259]: ok fine, as clarified, you're testing that the handler has not executed. this solution works then

File: docs/api/options.md [hunk 6]
```diff
@@ -89,6 +89,7 @@
 - `allowBatchedQueries`: Boolean. Flag to control whether to allow batched queries. When `true`, the server supports recieving an array of queries and returns an array of results.
 
 - `compilerOptions`: Object. Configurable options for the graphql-jit compiler. For more details check https://github.com/zalando-incubator/graphql-jit
+- `additionalRouteProps`: Object. Takes similar configuration to the Route Options of Fastify. Usecases include being able to add constraints, modify validation schema, increase the `bodyLimit`, etc. You can read more about the possible values on [fastify.dev's Routes Options](https://fastify.dev/docs/latest/Reference/Routes/#options)
```
Comment (simoneb) [inline docs/api/options.md:92]:
```suggestion
- `additionalRouteProps`: Object. Takes similar configuration to the Route Options of Fastify. Use cases include being able to add constraints, modify validation schema, increase the `bodyLimit`, etc. You can read more about the possible values on [fastify.dev's Routes Options](https://fastify.dev/docs/latest/Reference/Routes/#options)
```

I was also thinking, since the right term in the fastify docs is route options, why don't we call it options instead of props?

File: index.d.ts [hunk 7]
```diff
@@ -487,6 +488,18 @@ declare namespace mercurius {
...
+     * Customize the graphql routers initialized by mercurius with 
+     * more fastify route options. 
+     * 
+     * Usecases:
+     * - Add more validation 
+     * - change the schema structure 
+     * - Hook on the request's life cycle
+     * - Increase body size limit for larger queries
+    */
+    additionalRouteProps?:Omit<RouteOptions,"handler" | "wsHandler">
```
Comment (simoneb) [inline index.d.ts:502]: let's add types tests for this

File: docs/api/options.md (hunk 6 above)
Comment (barelyhuman) [reply docs/api/options.md:92]: Makes sense

File: index.d.ts (hunk 7 above)
Comment (barelyhuman) [reply index.d.ts:502]: There's no type tests for the options in place though, do we want it to be a separate PR for all of them ?

Comment (simoneb) [reply index.d.ts:502]: I mean, just add a test which checks that you can supply route options (and which ones perhaps) here https://github.com/mercurius-js/mercurius/blob/master/test/types/index.ts

**Review by simoneb — APPROVED** [review]

**Comment by simoneb** [comment]
@mcollina if you could take another look please

**Comment by barelyhuman** [comment]
@mcollina whenever you get time

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by simoneb** [comment]
Can somebody do a release? Speaking of which, is there an established way to do releases in this package (or the rest of the ecosystem)? If not, we shall consider introducing one @Eomm

**Comment by Eomm** [comment]
Cannot release - I'm not part of the `mercurius-js` org 🙌🏼

**Comment by simoneb** [comment]
@Eomm you should be then! In any case I was looping you in mostly to check if you had settled on anything with Fastify

**Comment by simoneb** [comment]
Can anybody do a release of mercurius? We are waiting for this feature to be released to start using it in one of our repos.
